lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -848,66 +1278,45 @@ def upload_reads(self, ctx, params):
         # ctx is the context object
         # return variables are: returnVal
         #BEGIN upload_reads
-        del ctx
         self.log('Starting upload reads, parsing args')
-        o, wsid, name, objid, kbtype, single_end, fwdid, revid, shock = (
-            self._proc_upload_reads_params(params))
-        # IF from shock fwdid and revid are shock nodes, if not shock they are file paths
+        o, wsid, name, objid, kbtype, single_end, fwdsource, revsource, reads_source = (
+                                                    self._proc_upload_reads_params(params))
+        # If reads_source == 'shock', fwdsource and revsource are shock nodes
+        # If reads_source == 'web', fwdsource and revsource are urls
+        # If reads_source == 'staging', fwdsource and revsource are file name in staging area
+        # If reads_source == 'local', fwdsource and revsource are file paths
         dfu = DataFileUtil(self.callback_url)
-        fwdpath = None
-        revpath = None
-        if shock:
-            # Grab files from Shock
-            fileinput = [{'shock_id': fwdid,
-                          'file_path': self.scratch + '/fwd/',
-                          'unpack': 'uncompress'}]
-            if revid:
-                fileinput.append({'shock_id': revid,
-                                  'file_path': self.scratch + '/rev/',
-                                  'unpack': 'uncompress'})
-            self.log('downloading reads file(s) from Shock')
-            files = dfu.shock_to_file_mass(fileinput)
-            fwdpath = files[0]["file_path"]
-            fwdname = files[0]["node_file_name"]
-            if revid:
-                revpath = files[1]["file_path"]
-                revname = files[1]["node_file_name"]
-        else:
-            # Not shock
-            fwdpath = fwdid
-            revpath = revid
-            fwdname = None
-            revname = None
-            fwdid = None
-            revid = None
+        fwdname, revname, fwdid, revid = (None,) * 4
+        ret = self._process_download(fwdsource, revsource, reads_source,
+                                     params.get('download_type'), ctx['user_id'])
+
+        fwdpath = ret.get('fwdpath')
+        revpath = ret.get('revpath')
+
+        if reads_source == 'shock':
+            fwdname = ret.get('fwdname')
+            revname = ret.get('revname')
+            fwdid = fwdsource
+            revid = revsource
 
         actualpath = fwdpath
         if revpath:
             # now interleave the files
-            actualpath = os.path.join(self.scratch, self.get_file_prefix() + '.inter.fastq')
+            actualpath = os.path.join(
+                self.scratch, self.get_file_prefix() + '.inter.fastq')
             self.interleave(None, None, fwdname, fwdid,
```

Comment:
@MrCreosote since I am logging all of my func activities, can I not touch this function? I am pretty sure users can find the url/staging file on the log.